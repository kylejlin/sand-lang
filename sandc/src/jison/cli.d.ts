export function main(opts: JisonCliOptions): void;

export interface JisonCliOptions {
  file: string;
  outfile?: string;
}
