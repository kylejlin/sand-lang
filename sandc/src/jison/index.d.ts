export class Parser {
  constructor(grammar: any);

  lexer: Scanner;
  yy: any;

  generate(): string;
  parse(program: string): any;
}

export function print(line: string): void;

export interface Scanner<TokenType = string> {
  yytext: string;
  yyleng: number;
  yylloc: JisonNodeLocation;
  /**
   * For whatever odd reason, Jison starts yylineno at 0,
   * so make sure to subtract one if your line count is 1-based.
   */
  yylineno: number;

  lex(): TokenType;
  setInput(input: string): void;

  pastInput(): string;
  upcomingInput(): string;
  input(): void;
}

export interface JisonNodeLocation {
  first_line: number;
  last_line: number;
  first_column: number;
  last_column: number;
}
