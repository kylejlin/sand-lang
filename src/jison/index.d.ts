export class Parser {
  constructor(grammar: any);

  parser: { yy: any };

  generate(): string;
  parse(program: string): any;
}

export function print(line: string): void;
