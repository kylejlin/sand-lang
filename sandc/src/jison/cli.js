const jisonCli = require("jison/lib/cli");

module.exports = jisonCli;
