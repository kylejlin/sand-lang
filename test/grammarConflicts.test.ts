import jison from "./jison";
import fs from "fs";
import path from "path";

const GRAMMAR_PATH = path.join(__dirname, "../grammar/sand.jison");

test("sand.jison does not have new conflicts", () => {
  let out = "";
  function print(line: string) {
    out += line;
    out += "\n";
  }

  jison.print = print;

  const src = fs.readFileSync(GRAMMAR_PATH, "utf8");
  const _parser = new jison.Parser(src);

  expect(out).toMatchSnapshot();

  const conflictCount = getConflictCount(out);

  expect(conflictCount).toMatchSnapshot();
});

function getConflictCount(logs: string): ConflictCount {
  const matches = logs.match(
    /Conflict in grammar[\s\S]*?(?=(?:Conflict in grammar|States with conflicts|$))/g,
  );

  if (matches === null) {
    return { total: 0, reduceShift: 0, reduceReduce: 0 };
  } else {
    return {
      total: matches.length,
      reduceShift: matches.filter(isReduceShift).length,
      reduceReduce: matches.filter(isReduceReduce).length,
    };
  }
}

function isReduceShift(match: string): boolean {
  return /- reduce[\s\S]*- shift/.test(match);
}

function isReduceReduce(match: string): boolean {
  return /- reduce[\s\S]*- reduce/.test(match);
}

interface ConflictCount {
  total: number;
  reduceShift: number;
  reduceReduce: number;
}
