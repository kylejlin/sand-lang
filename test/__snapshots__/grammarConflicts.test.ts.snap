// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sand.jison does not have new conflicts 1`] = `
"Conflict in grammar: multiple actions possible when lookahead token is || in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 124)
Conflict in grammar: multiple actions possible when lookahead token is && in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 125)
Conflict in grammar: multiple actions possible when lookahead token is == in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 126)
Conflict in grammar: multiple actions possible when lookahead token is != in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 127)
Conflict in grammar: multiple actions possible when lookahead token is < in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 128)
Conflict in grammar: multiple actions possible when lookahead token is <= in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 129)
Conflict in grammar: multiple actions possible when lookahead token is > in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 130)
Conflict in grammar: multiple actions possible when lookahead token is >= in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 131)
Conflict in grammar: multiple actions possible when lookahead token is + in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 132)
Conflict in grammar: multiple actions possible when lookahead token is - in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 133)
Conflict in grammar: multiple actions possible when lookahead token is * in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 134)
Conflict in grammar: multiple actions possible when lookahead token is / in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 135)
Conflict in grammar: multiple actions possible when lookahead token is % in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 136)
Conflict in grammar: multiple actions possible when lookahead token is ** in state 76
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken
- shift token (then go to state 137)
Conflict in grammar: multiple actions possible when lookahead token is || in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 138)
Conflict in grammar: multiple actions possible when lookahead token is && in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 139)
Conflict in grammar: multiple actions possible when lookahead token is == in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 140)
Conflict in grammar: multiple actions possible when lookahead token is != in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 141)
Conflict in grammar: multiple actions possible when lookahead token is < in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 142)
Conflict in grammar: multiple actions possible when lookahead token is <= in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 143)
Conflict in grammar: multiple actions possible when lookahead token is > in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 144)
Conflict in grammar: multiple actions possible when lookahead token is >= in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 145)
Conflict in grammar: multiple actions possible when lookahead token is + in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 146)
Conflict in grammar: multiple actions possible when lookahead token is - in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 147)
Conflict in grammar: multiple actions possible when lookahead token is * in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 148)
Conflict in grammar: multiple actions possible when lookahead token is / in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 149)
Conflict in grammar: multiple actions possible when lookahead token is % in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 150)
Conflict in grammar: multiple actions possible when lookahead token is ** in state 77
- reduce by rule: expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken
- shift token (then go to state 151)
Conflict in grammar: multiple actions possible when lookahead token is ** in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 113)
Conflict in grammar: multiple actions possible when lookahead token is % in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 112)
Conflict in grammar: multiple actions possible when lookahead token is / in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 111)
Conflict in grammar: multiple actions possible when lookahead token is * in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 110)
Conflict in grammar: multiple actions possible when lookahead token is - in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 109)
Conflict in grammar: multiple actions possible when lookahead token is + in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 108)
Conflict in grammar: multiple actions possible when lookahead token is >= in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 107)
Conflict in grammar: multiple actions possible when lookahead token is > in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 106)
Conflict in grammar: multiple actions possible when lookahead token is <= in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 105)
Conflict in grammar: multiple actions possible when lookahead token is < in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 104)
Conflict in grammar: multiple actions possible when lookahead token is != in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 103)
Conflict in grammar: multiple actions possible when lookahead token is == in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 102)
Conflict in grammar: multiple actions possible when lookahead token is && in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 101)
Conflict in grammar: multiple actions possible when lookahead token is || in state 153
- reduce by rule: expression -> expressionIncludingRightDelimiter
- shift token (then go to state 100)

States with conflicts:
State 76
  expressionLackingRightDelimiter -> expressionLackingRightDelimiterStartingWithInfixToken . #lookaheads= } ; { || && == != < <= > >= + - * / % ** , ] )
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .|| expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .&& expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .== expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .!= expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .< expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .<= expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .> expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .>= expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .+ expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .- expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .* expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken ./ expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .% expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterStartingWithInfixToken -> expressionLackingRightDelimiterStartingWithInfixToken .** expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
State 77
  expressionLackingRightDelimiter -> expressionLackingRightDelimiterNotStartingWithInfixToken . #lookaheads= } ; { || && == != < <= > >= + - * / % ** , ] )
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .|| expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .&& expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .== expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .!= expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .< expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .<= expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .> expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .>= expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .+ expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .- expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .* expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken ./ expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .% expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionLackingRightDelimiterNotStartingWithInfixToken .** expression #lookaheads= ; } || && == != < <= > >= + - * / % ** { ] ) ,
State 153
  expression -> expressionIncludingRightDelimiter . #lookaheads= { ** % / * - + >= > <= < != == && || } ; ] ) ,
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .|| expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .&& expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .== expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .!= expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .< expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .<= expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .> expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .>= expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .+ expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .- expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .* expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter ./ expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .% expression
  expressionLackingRightDelimiterNotStartingWithInfixToken -> expressionIncludingRightDelimiter .** expression
"
`;

exports[`sand.jison does not have new conflicts 2`] = `
Object {
  "reduceReduce": 0,
  "reduceShift": 42,
  "total": 42,
}
`;
